### PR TITLE
[ALBEF] model refactoring

### PR DESCRIPTION
**Pull request by katrina433**
Summary:
This PR refactors `ALBEFModel` for different downstream tasks. 
* `ALBEFModel`: the core model that outputs unimodal and multimodal embeddings (momentum embeddings too). 
(VQA can use this class because it only needs mlm loss)

* `ALBEFModelWithSimilarity`: outputs unimodal and multimodal embeddings (no momentum embeddings), negative image-text multimodal embeddings (for ITM loss), and similarity outputs (for ITC loss). 
(Retrieval can use this class because it needs ITM and ITC losses)

Test plan:
Passed unit tests in `test/models/test_albef.py`

### CODE REVIEW

File: torchmultimodal/models/albef.py [hunk 1]
```diff
@@ -262,25 +307,27 @@ def _neg_embeddings(
...
+
+
+def _gather_embeddings(embeddings: torch.Tensor) -> torch.Tensor:
+    if not torch.distributed.is_available() or not torch.distributed.is_initialized():
+        return embeddings
+
+    embeddings_all_gpus = [
+        torch.zeros_like(embeddings) for _ in range(torch.distributed.get_world_size())
+    ]
+    torch.distributed.all_gather(embeddings_all_gpus, embeddings, async_op=False)
```
Comment (ebsmothers) [inline torchmultimodal/models/albef.py:331]: I think `async_op=False` should be the default value here?

File: torchmultimodal/models/albef.py [hunk 2]
```diff
@@ -124,74 +185,47 @@ def forward(
...
-        )
-        image_embeds_m, image_feat_m, text_feat_m = self._momentum_embeddings(
-            image, text, text_atts
+        idx: Tensor,
+    ) -> ALBEFWithSimilarityOutput:
+        outputs = self.model(image, text, text_atts)
+
+        idx = idx.view(-1, 1)
+        idx_all = torch.cat([idx.t(), self.idx_queue.detach().clone()], dim=1)
+        pos_idx = torch.eq(idx, idx_all).float()
```
Comment (ebsmothers) [inline torchmultimodal/models/albef.py:194]: Add some comments explaining this portion

File: torchmultimodal/models/albef.py [hunk 3]
```diff
@@ -262,25 +311,27 @@ def _neg_embeddings(
...
+
+@torch.no_grad()
+def _copy_params_momentum_models(model: nn.Module, model_m: nn.Module) -> None:
+    for param, param_m in zip(model.parameters(), model_m.parameters()):
+        param_m.data.copy_(param.data)
+        param_m.requires_grad = False
+
+
+@torch.no_grad()
+def _momentum_update(model: nn.Module, model_m: nn.Module, momentum: float) -> None:
```
[the remark this answers is not in the document]
Comment (ebsmothers) [reply torchmultimodal/models/albef.py:323]: Another option here is to make a mixin class for momentum models. That's a bit more involved so probably not necessary for this PR, but agree with moving this to a util.

**Comment by ebsmothers** [comment]
Re @YosuaMichael's point about inference mode, an alternative option to the if/else approach you describe is to create a separate bare-bones class containing only the non-momentum models, e.g. `ALBEFForInference`. But then there are a couple follow-up questions:

(1) What does the forward look like? It will  depend on the task a bit more, but perhaps we could just return image, text, and multimodal embeddings. Then let the specific task handle the remaining pieces (similarity for retrieval and text decoder for VQA).
(2) In a training and eval loop, do we then use a separate class for each of these? That could be clunky (we'd have to do lots of param copying between the two).

File: torchmultimodal/models/albef.py [hunk 4]
```diff
@@ -40,81 +37,145 @@
...
         self.vision_encoder = vision_encoder
         self.text_encoder = text_encoder
         self.multimodal_encoder = multimodal_encoder
-        self.vision_proj = vision_proj
-        self.text_proj = text_proj
         self.vision_encoder_m = copy.deepcopy(vision_encoder)
         self.text_encoder_m = copy.deepcopy(text_encoder)
         self.multimodal_encoder_m = copy.deepcopy(multimodal_encoder)
+
+        _copy_params_momentum_models(self.vision_encoder, self.vision_encoder_m)
```
[the remark this answers is not in the document]
Comment (YosuaMichael) [reply torchmultimodal/models/albef.py]: I see, in that case, I think it will be simpler to just call `self.vision_encoder_m.eval()` in that case

**Comment by YosuaMichael** [comment]
> Re @YosuaMichael's point about inference mode, an alternative option to the if/else approach you describe is to create a separate bare-bones class containing only the non-momentum models, e.g. `ALBEFForInference`. But then there are a couple follow-up questions:
> 
> (1) What does the forward look like? It will depend on the task a bit more, but perhaps we could just return image, text, and multimodal embeddings. Then let the specific task handle the remaining pieces (similarity for retrieval and text decoder for VQA). (2) In a training and eval loop, do we then use a separate class for each of these? That could be clunky (we'd have to do lots of param copying between the two).

I agree, I think this is also a good approach. From @ebsmothers idea, here is what I think:
```
class AlbefBaseModel(nn.Module):
    def __init__(
        self,
        vision_encoder: nn.Module,
        text_encoder: nn.Module,
        multimodal_encoder: nn.Module,
    ) -> None:
        super().__init__()
        self.vision_encoder = vision_encoder
        self.text_encoder = text_encoder
        self.multimodal_encoder = multimodal_encoder

    def forward(
        self,
        image: Tensor,
        text: Tensor,
        text_atts: Tensor,
    ) -> ALBEFOutput:
        image_embeds = self.vision_encoder(image)
        text_embeds = self.text_encoder(text, text_atts)
        multimodal_embeddings = self.multimodal_encoder(
            image_embeds, text_embeds, text_atts
        )

        return ALBEFBaseOutput(
            image_embeddings=image_embeds,
            text_embeddings=text_embeds,
            multimodal_embeddings=multimodal_embeddings,
        )

class AlbefModel(nn.Module):
    def __init__(
        self,
        base_model: nn.Module,
        momentum: float
    ) -> None:
        super().__init__()
        self.model = base_model
        self.model_m = copy.deepcopy(base_model).eval()
        self.momentum = momentum

  def forward(
        self,
        image: Tensor,
        text: Tensor,
        text_atts: Tensor,
    ) -> ALBEFOutput:
        momentum_update(self.model, self.model_m, self.momentum)
        base_output = self.model(image, text, text_atts)
        momentum_output = self.model_m(image, text, text_atts)
        return ALBEFOutput(
            base_output=base_output,
            momentum_output=momentum_output,
        )
```

**Review by YosuaMichael — APPROVED** [review]
For the further refactoring on the momentum (to enable having model without momentum for inference), after talking with @katrina433 , I agree that we might want to have it while developing the train and inference loop.

In this case, I will approve current refactoring effort first so it is not blocking.
